src/main/java/io/odpf/firehose/sink/objectstorage/writer/WriterOrchestrator.java

```diff
@@ -102,11 +111,26 @@ private void checkStatus() throws Exception {
      */
     public String write(Record record) throws Exception {
         checkStatus();
-        Path partitionedPath = localStorage.getTimePartitionPath().create(record);
-        String pathString = partitionedPath.toString();
+        Partition partition = localStorage.getPartitionFactory().getPartition(record);
+
+        String dateTimePartition = partition.getDatetimePathWithoutPrefix();
+
         synchronized (timePartitionWriterMap) {
```

Comment:
The synchronization is too costly, we can have a check before hand & synchronize only when there is a need to create the writer. & synchronization block can be just limited at the time of creation of the writer.

Reply:
we are doing computeifabsent() which is not atomic. So synchronised is required.

Reply:
let me think this again as we are using ConcurrentHashMap, maybe we can get away without synchronised block.

Reply:
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#computeIfAbsent-K-java.util.function.Function-
this is atomic :)

Reply:
although concurrent hashmap `remove()` is also atomic, tricky part are on the writing and closing writer

there might still data race problem, writer might be acquired right before key removal but we cant ensure closing is happened after the writing record is finished

we might be able optimise by apply synchronisation only on writer access, need thinking through on the implementation

Reply:
adding lock on the writer will solve this, but on the current implementation removing synchronisation will not increase the throughput, because only single record are written at a time, 
we can increase the throughput further by writing records in parallel for each partition

Reply:
we are gonna fix this in another PR